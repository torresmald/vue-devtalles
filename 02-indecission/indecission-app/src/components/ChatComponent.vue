<script setup lang="ts">
import type { ChatMessages } from '@/interfaces/chat.interface';
import ChatBubble from './ChatBubble.vue';
import { ref, watch } from 'vue';



interface Props {
  messages: ChatMessages[];
}
const props = defineProps<Props>();

const messagesRef = ref<ChatMessages[]>(props.messages);
  const chatRef = ref<HTMLDivElement | null>(null);

watch( messagesRef.value, () => {
  console.log('scrolling')
  chatRef.value?.scrollTo({
    top: chatRef.value.scrollHeight,
    behavior: 'smooth'
  })
})
</script>

<!-- Fuente: https://tailwindcomponents.com/component/chat-layout -->
<template>
  <div ref="chatRef" class="flex-1 overflow-y-auto p-4">
    <div class="flex flex-col space-y-2">
      <ChatBubble v-for="message in messages" :key="message.id" v-bind="message" />
    </div>
  </div>
</template>
