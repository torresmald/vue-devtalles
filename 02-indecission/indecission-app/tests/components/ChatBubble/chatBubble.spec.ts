import { describe } from 'node:test';
import { expect, test, vi } from 'vitest';
import ChatComponent from '@/components/ChatComponent.vue';
import { mount } from '@vue/test-utils';
import type { ChatMessages } from '@/interfaces/chat.interface';

const messages: ChatMessages[] = [
  {
    isMine: true,
    message: 'Hello',
  },
  {
    isMine: false,
    message: 'Hi',
    image: 'https://via.placeholder.com/150',
  },
];

describe('<ChatBubble />', () => {
  const wrapper = mount(ChatComponent, {
    props: { messages: messages },
  });

  test('should render the chat bubble', () => {
    expect(wrapper.html()).toMatchSnapshot();

    const chatComponent = wrapper.findAllComponents({ name: 'ChatBubble' });

    expect(chatComponent.length).toBe(messages.length);
  });

  // test('scroll component', async () => {
  //   const scrollMock = vi.fn();

  //   const chatRef = wrapper.vm.$refs.chatRef as HTMLDivElement;

  //   chatRef.scrollTo = scrollMock;

  //   await wrapper.setProps({ messages: [...messages, { message: 'Hey mas', isMine: true }] });

  //   await new Promise((resolve) => setTimeout(resolve, 300));

  //   expect(scrollMock).toHaveBeenCalledWith({
  //     top: chatRef.scrollHeight,
  //     behavior: 'smooth',
  //   });
  // });
});
