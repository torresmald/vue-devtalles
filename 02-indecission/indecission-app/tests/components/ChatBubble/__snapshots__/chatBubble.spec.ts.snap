// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render the chat bubble 1`] = `
"<div class="flex-1 overflow-y-auto p-4">
  <div class="flex flex-col space-y-2">
    <!-- Example Message -->
    <div class="flex justify-end">
      <div class="bg-blue-200 text-black p-2 rounded-lg max-w-xs">Hello</div>
    </div><!-- Example Message -->
    <!-- Example Received Message -->
    <div class="flex">
      <div class="bg-gray-300 text-black p-2 rounded-lg max-w-xs"><span class="capitalize">Hi</span><img src="https://via.placeholder.com/150" alt="Image" class="w-52 h-52 rounded-md object-cover"></div>
    </div>
  </div>
</div>"
`;
